google-cloud-spanner/src/main/java/com/google/cloud/spanner/GrpcResultSet.java

```diff
@@ -49,6 +50,18 @@ class GrpcResultSet extends AbstractResultSet<List<Object>> {
     this.decodeMode = decodeMode;
   }
 
+  @Override
+  public boolean canGetProtobufValue(int columnIndex) {
+    return !closed && currRow != null && currRow.canGetProtoValue(columnIndex);
+  }
+
+  @Override
+  public Value getProtobufValue(int columnIndex) {
+    checkState(!closed, "ResultSet is closed");
+    checkState(currRow != null, "next() call required");
+    return currRow.getProtoValueInternal(columnIndex);
```

Comment:
Please correct me if I am wrong but IMO the idea of this function is to directly get the protobuf values without conversion to java object but I think we are first converting proto value to a java object and this function then converts the java object back to proto

Reply:
Your understanding of the intention here is correct. We do this in order to get the protobuf values directly without conversion.

The method however does get the protobuf value directly (or it fails if it is no longer available). `currRow` contains a list of the protobuf values in the following cases:
1. Decoding has been set to LAZY_PER_ROW and none of the columns have so far been decoded. The first time one column is decoded, all columns are decoded, and this method can no longer be called.
2. Decoding has been set to LAZY_PER_COL and this specific column has not yet been decoded.

If decoding is set to DIRECT, then this method will always fail. The `currRow.getProtovalueInternal` call will check that the protobuf value is still actually available.